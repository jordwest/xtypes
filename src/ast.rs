use std::collections::BTreeMap;

/// An attribute is a special flag that can be attached to:
///  - A message
///  - A field
///  - An enum variant
///  - A module
///  - A use statement
///
/// Attributes can be used by the code generator in any way
/// they're needed.
///
/// For example, these attributes might tell a code
/// generator to name the fields differently when
/// doing JSON serialization:
///
/// ```xt
/// message Book = {
///   #[json.key="book_name"]
///   name: string,
///
///   #[json.key="author_name"]
///   author: string,
/// }
/// ```
#[derive(Clone, Debug, PartialEq)]
pub struct Attribute {
    pub name: String,
    pub value: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct AttributeList(BTreeMap<String, Attribute>);

impl AttributeList {
    pub fn new() -> Self {
        AttributeList(BTreeMap::new())
    }

    pub fn add(&mut self, attr: Attribute) {
        self.0.insert(attr.name.clone(), attr);
    }

    pub fn get<T: AsRef<str> + Sized>(&self, key: T) -> Option<String> {
        self.0.get(key.as_ref()).map(|a| a.value.clone())
    }
}

/// A tuple can contain multiple types in a sequence.
///
/// For example:
///
/// ```xt
/// (string, float32)
/// ```
///
/// This is useful in the context of an enum:
///
/// ```xt
/// message OneOrTwoNumbers =
///     | One(float32)
///     | Two(float32, float32)
/// ;
/// ```
#[derive(Clone, Debug, PartialEq)]
pub struct Tuple(pub Vec<String>);

#[derive(Clone, Debug, PartialEq)]
pub struct EnumMessage {
    pub variants: Vec<EnumVariant>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EnumVariant {
    pub name: String,
    pub attrs: AttributeList,
    pub content: Option<Tuple>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructField {
    pub name: String,
    pub type_name: TypeName,
    pub is_optional: bool,
    pub attrs: AttributeList,
}
#[derive(Clone, Debug, PartialEq)]
pub struct StructMessage {
    pub fields: Vec<StructField>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum MessageType {
    Enum(EnumMessage),
    Struct(StructMessage),
}

#[derive(Clone, Debug, PartialEq)]
pub enum SymbolType {
    Message(MessageType),
    Primitive,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SymbolDefinition {
    pub name: TypeName,
    pub attrs: AttributeList,
    pub value: SymbolType,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ModuleInfo {
    pub name: String,
    pub attrs: AttributeList,
}

/// A portion of a [DottedIdent](xtypes::ast::DottedIdent).
#[derive(Clone, Debug, PartialEq)]
pub enum DottedIdentPart {
    Ident(String),
    Wildcard,
}

/// A DottedIdent refers to a dotted variable name. For example:
///
/// ```xt
/// use SomeModule.SubModule.*;
/// //  ^^^^^^^^^^^^^^^^^^^^^^
/// ```
///
/// The above statement would contain 3 "parts":
///   - Ident("SomeModule")
///   - Ident("SubModule")
///   - Wildcard
#[derive(Clone, Debug, PartialEq)]
pub struct DottedIdent {
    pub parts: Vec<DottedIdentPart>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum IdentOrWildcard {
    Ident(String),
    Wildcard,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ModuleUse {
    pub attrs: AttributeList,
    pub filename: String,
    pub ident: IdentOrWildcard,
}

#[derive(Clone, Debug, PartialEq)]
pub enum TypeName {
    Concrete(String),
    Generic(String, Box<TypeName>),
}
impl Into<String> for &TypeName {
    fn into(self) -> String {
        match &self {
            TypeName::Concrete(s) => s.clone(),
            TypeName::Generic(s, _) => s.clone(),
        }
    }
}
impl Into<String> for TypeName {
    fn into(self) -> String {
        (&self).into()
    }
}
impl TypeName {
    pub fn identifier(&self) -> String {
        match self {
            TypeName::Concrete(s) => s.clone(),
            TypeName::Generic(s, _) => s.clone(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct XtFile {
    pub module_info: ModuleInfo,
    pub use_imports: Vec<ModuleUse>,
    pub symbols: Vec<SymbolDefinition>,
}
