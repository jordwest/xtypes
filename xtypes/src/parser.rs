use crate::ast::*;
use pest::iterators::Pair;
use pest::Parser;
use pest_derive::Parser;

/// The parser takes a `.xt` file and parses it into an AST

#[derive(Parser)]
#[grammar = "grammar.pest"]
pub struct XtParser;

impl From<Pair<'_, Rule>> for Attribute {
    fn from(pair: Pair<'_, Rule>) -> Attribute {
        match pair.as_rule() {
            Rule::attribute => {
                let mut iter = pair.into_inner().into_iter();
                Attribute {
                    name: iter.next().unwrap().as_str().into(),
                    value: iter.next().unwrap().as_str().into(),
                }
            }
            _ => panic!(),
        }
    }
}

impl From<Pair<'_, Rule>> for Tuple {
    fn from(pair: Pair<'_, Rule>) -> Tuple {
        let mut types = vec![];
        for pair in pair.into_inner() {
            match pair.as_rule() {
                Rule::dotted_ident => types.push(pair.as_str().into()),
                _ => panic!(),
            }
        }
        Tuple(types)
    }
}
impl From<Pair<'_, Rule>> for EnumVariant {
    fn from(pair: Pair<'_, Rule>) -> EnumVariant {
        match pair.as_rule() {
            Rule::variant => {
                let mut name = String::new();
                let mut attrs = AttributeList::new();
                let mut content = None;
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::ident => name = pair.as_str().into(),
                        Rule::attribute => attrs.add(pair.into()),
                        Rule::tuple => content = Some(pair.into()),
                        _ => panic!(),
                    }
                }
                EnumVariant {
                    name,
                    attrs,
                    content,
                }
            }
            _ => panic!(),
        }
    }
}

impl From<Pair<'_, Rule>> for EnumMessage {
    fn from(pair: Pair<'_, Rule>) -> EnumMessage {
        match pair.as_rule() {
            Rule::enum_message => EnumMessage {
                variants: pair.into_inner().into_iter().map(|i| i.into()).collect(),
            },
            _ => panic!(),
        }
    }
}

impl From<Pair<'_, Rule>> for TypeName {
    fn from(pair: Pair<'_, Rule>) -> TypeName {
        let pair = pair.into_inner().next().unwrap();
        match pair.as_rule() {
            Rule::dotted_ident => TypeName::Concrete(pair.as_str().into()),
            Rule::generic_type => {
                let mut outside_type = None;
                let mut inside_type = None;
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::ident => outside_type = Some(pair.as_str().into()),
                        Rule::type_name => inside_type = Some(pair.into()),
                        _ => panic!(),
                    }
                }
                TypeName::Generic(outside_type.unwrap(), Box::new(inside_type.unwrap()))
            }
            _ => panic!(),
        }
    }
}
impl From<Pair<'_, Rule>> for StructField {
    fn from(pair: Pair<'_, Rule>) -> StructField {
        match pair.as_rule() {
            Rule::struct_field => {
                let mut name = None;
                let mut type_name = None;
                let mut is_optional = false;
                let mut attrs = AttributeList::new();
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::field_name => name = Some(pair.as_str().into()),
                        Rule::type_name => type_name = Some(pair.into()),
                        Rule::attribute => attrs.add(pair.into()),
                        Rule::optional => is_optional = true,
                        _ => panic!("Unexpected rule {}", pair.as_str()),
                    }
                }
                StructField {
                    name: name.unwrap(),
                    type_name: type_name.unwrap(),
                    is_optional,
                    attrs,
                }
            }
            _ => panic!(),
        }
    }
}

impl From<Pair<'_, Rule>> for StructMessage {
    fn from(pair: Pair<'_, Rule>) -> StructMessage {
        match pair.as_rule() {
            Rule::struct_message => StructMessage {
                fields: pair.into_inner().into_iter().map(|i| i.into()).collect(),
            },
            _ => panic!(),
        }
    }
}
impl From<Pair<'_, Rule>> for MessageType {
    fn from(pair: Pair<'_, Rule>) -> MessageType {
        match pair.as_rule() {
            Rule::enum_message => MessageType::Enum(pair.into()),
            Rule::struct_message => MessageType::Struct(pair.into()),
            r => panic!("Unexpected message type {:?}", r),
        }
    }
}

impl From<Pair<'_, Rule>> for SymbolDefinition {
    fn from(pair: Pair<'_, Rule>) -> SymbolDefinition {
        let mut name: Option<TypeName> = None;
        let mut value: Option<MessageType> = None;
        let mut attrs = AttributeList::new();
        let rule = pair.as_rule();
        for pair in pair.into_inner() {
            match pair.as_rule() {
                Rule::type_name => name = Some(pair.into()),
                Rule::attribute => attrs.add(pair.into()),
                _ => value = Some(pair.into()),
            }
        }
        match rule {
            Rule::typedef => SymbolDefinition {
                name: name.unwrap().into(),
                attrs,
                value: SymbolType::Primitive,
            },
            Rule::message => SymbolDefinition {
                name: name.unwrap().into(),
                attrs,
                value: SymbolType::Message(value.unwrap()),
            },
            unknown => panic!("Unexpected rule '{:?}' found ", unknown),
        }
    }
}
impl From<Pair<'_, Rule>> for ModuleInfo {
    fn from(pair: Pair<'_, Rule>) -> ModuleInfo {
        match pair.as_rule() {
            Rule::module_decl => {
                let mut attrs = AttributeList::new();
                let mut name = "";
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::dotted_ident => name = pair.as_str(),
                        Rule::attribute => attrs.add(pair.into()),
                        _ => panic!(),
                    }
                }
                ModuleInfo {
                    name: name.into(),
                    attrs,
                }
            }
            _ => panic!(),
        }
    }
}

impl From<Pair<'_, Rule>> for DottedIdent {
    fn from(pair: Pair<'_, Rule>) -> DottedIdent {
        match pair.as_rule() {
            Rule::dotted_ident => {
                let mut parts = vec![];
                println!("Outer: {}", pair.as_str());
                for pair in pair.into_inner() {
                    println!("Pair: {}", pair.as_str());
                    match pair.as_rule() {
                        Rule::ident => parts.push(DottedIdentPart::Ident(pair.as_str().into())),
                        Rule::wildcard => parts.push(DottedIdentPart::Wildcard),
                        _ => panic!("Unexpected rule"),
                    }
                }
                DottedIdent { parts }
            }
            unknown => panic!("Unexpected rule '{:?}' found ", unknown),
        }
    }
}

impl From<Pair<'_, Rule>> for IdentOrWildcard {
    fn from(pair: Pair<'_, Rule>) -> IdentOrWildcard {
        match pair.as_rule() {
            Rule::ident => IdentOrWildcard::Ident(pair.as_str().to_owned()),
            Rule::wildcard => IdentOrWildcard::Wildcard,
            r => panic!("Unexpected rule {:?}", r),
        }
    }
}

impl From<Pair<'_, Rule>> for ModuleUse {
    fn from(pair: Pair<'_, Rule>) -> ModuleUse {
        match pair.as_rule() {
            Rule::use_statement => {
                let mut ident = None;
                let mut filename = None;
                let mut attrs = AttributeList::new();
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::ident => {
                            ident = Some(IdentOrWildcard::Ident(pair.as_str().to_owned()))
                        }
                        Rule::wildcard => ident = Some(IdentOrWildcard::Wildcard),
                        Rule::filename => filename = Some(pair.as_str()),
                        Rule::attribute => attrs.add(pair.into()),
                        r => panic!("Unexpected rule {:?}", r),
                    }
                }
                ModuleUse {
                    attrs,
                    ident: ident.unwrap(),
                    filename: filename.unwrap().to_owned(),
                }
            }
            unknown => panic!("Unexpected rule '{:?}' found ", unknown),
        }
    }
}

impl From<Pair<'_, Rule>> for XtFile {
    fn from(pair: Pair<'_, Rule>) -> XtFile {
        match pair.as_rule() {
            Rule::file => {
                let mut symbols = vec![];
                let mut use_imports = vec![];
                let mut module_info = None;
                for pair in pair.into_inner() {
                    match pair.as_rule() {
                        Rule::module_decl => module_info = Some(pair.into()),
                        Rule::message => symbols.push(pair.into()),
                        Rule::use_statement => use_imports.push(pair.into()),
                        Rule::typedef => symbols.push(pair.into()),
                        Rule::EOI => (),
                        _ => panic!("Unexpected '{:?}'", pair),
                    }
                }
                XtFile {
                    module_info: module_info.unwrap(),
                    symbols,
                    use_imports,
                }
            }
            _ => panic!(),
        }
    }
}

pub fn parse(t: &str) -> XtFile {
    match XtParser::parse(Rule::file, t) {
        Err(e) => panic!("{}", e),
        Ok(v) => v.into_iter().next().unwrap().into(),
    }
}

#[test]
fn test_parse_sample_1() {
    use insta::assert_debug_snapshot_matches;

    let file = parse(include_str!("./sample.xt"));
    assert_debug_snapshot_matches!("sample.xt", file);
}
