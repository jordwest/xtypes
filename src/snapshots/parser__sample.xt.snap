---
created: "2019-02-10T11:45:48.623507Z"
creator: insta@0.6.2
source: src/parser.rs
expression: file
---
File {
    module_info: ModuleInfo {
        name: "Sample.Test",
        attrs: [
            Attribute {
                name: "output",
                value: "/sample/test/somefile.txt"
            }
        ]
    },
    messages: [
        Message {
            name: "EnumSample",
            attrs: [],
            value: Enum(
                EnumMessage {
                    variants: [
                        EnumVariant {
                            name: "One",
                            attrs: [
                                Attribute {
                                    name: "thing.stuff",
                                    value: "xyz"
                                },
                                Attribute {
                                    name: "json.tag",
                                    value: "A"
                                }
                            ],
                            content: None
                        },
                        EnumVariant {
                            name: "Two",
                            attrs: [
                                Attribute {
                                    name: "blah",
                                    value: "X"
                                }
                            ],
                            content: Some(
                                Tuple(
                                    [
                                        "StructSample",
                                        "Boolean"
                                    ]
                                )
                            )
                        },
                        EnumVariant {
                            name: "Three",
                            attrs: [],
                            content: None
                        }
                    ]
                }
            )
        },
        Message {
            name: "StructSample",
            attrs: [
                Attribute {
                    name: "http.method",
                    value: "GET"
                }
            ],
            value: Struct(
                StructMessage {
                    fields: [
                        StructField {
                            name: "repeated_field",
                            type_name: Generic(
                                "Thing",
                                Generic(
                                    "Array",
                                    Concrete(
                                        "Type"
                                    )
                                )
                            ),
                            is_optional: false,
                            attrs: [
                                Attribute {
                                    name: "json.key",
                                    value: "X"
                                }
                            ]
                        },
                        StructField {
                            name: "optional_field",
                            type_name: Concrete(
                                "type"
                            ),
                            is_optional: true,
                            attrs: [
                                Attribute {
                                    name: "json.key",
                                    value: "Y"
                                }
                            ]
                        }
                    ]
                }
            )
        }
    ]
}
